### ISSUE TITLE
Make decoding from a root key easier

### ISSUE THREAD

**Issue by tonyd256** [thoughtbot/Argo#141]
Currently to decode something in a root key like this:
```
{
  "user":
  {
    "id": 5,
    "name": "Bob"
  }
}
```
You need to do extra work to pull out the `User` object from the root key like so:
```swift
let json: AnyObject? = NSJSONSerialization.JSONObjectWithData ...
let user: User? = json >>- decode >>- { ($0 <| "user").value }
let decodedUser: Decoded<User> = json >>- decode >>- { $0 <| "user" } 
```

The purpose of the `decode` function was to make it easy to decode the JSON into the model without having to worry about the operators or using different code to go between a `Decoded` or `Optional` return type.

However, this breaks down with root keys. I think this feels weird and ruins the purpose of the `decode` function. How can we fix this? I have 3 propositions.

Overload all the `decode` operators to take a root key making the API look like so:
```swift
let user: User? = json >>- decode(rootKey: "user") // with root key
let user: User? = json >>- decode // without root key
```
I think this is super easy for the user to type, but adds a lot of code to the library.

Add the rootKey as a parameter on all decode functions:
```swift
let user: User? = json >>- decode(rootKey: "user") // with root key
let user: User? = json >>- decode(rootKey: .None) // without root key
```
I think this is not too much more work for the user but is a little weird. Could also make the `rootKey` label not required but then I think that might look confusing.

Add a new function for pulling out rootKeys:
```swift
let user: User? = json >>- rootKey("user") >>- decode // with root key
let user: User? = json >>- decode // without root key
```
This is also not so much work for the user and things don't change if you don't have root keys (like the first option). It also adds minimal code to the repo. I imagine it would look like this:
```swift
public func rootKey(key: String)(object: AnyObject) -> AnyObject? {
  return (object as? [String: AnyObject])?[key]
}
```

I'm also open to other suggestions.

**Comment by alskipp** [comment]
Do you think the following could be a possible solution?
```swift
func decode<T : Decodable where T == T>(rootKey: String? = .None)(object: AnyObject) -> T?
```
```swift
let user: User? = json >>- decode(rootKey: "user") // with root key
let user: User? = json >>- decode // without root key
```

**Comment by tonyd256** [comment]
@alskipp Let me know what you find ... @gfontenot suggested this too but I found it not possible.

**Comment by alskipp** [comment]
I'm not having any luck getting that idea to work. Doesn't seem possible currently, maybe it will be next week with Swift 2.0?

**Comment by alskipp** [comment]
With a bit more experimenting, this might be possible, but would require calling `decode` with empty parentheses for the case without a root key:

```swift
let user: User? = json >>- decode(rootKey: "user") // with root key
let user: User? = json >>- decode() // without root key
```

Not sure if this would be a good idea?

Here's a rough proof of concept:
```swift
infix operator >>- { associativity left precedence 100 }
func >>- <A,B>(x:A?, f:A -> B?) -> B? {
  return map(x) { f($0) } ?? .None
}

func f(x:Int? = .None)(y:Int) -> Int? {
  return map(x) { $0 + y }
}

Optional(1) >>- f() // .None
Optional(1) >>- f(x: 2) // .Some(3)
```

**Comment by tonyd256** [comment]
@alskipp yup, that's what I was getting at with option 2. Although, I wrote out the param and mentioned in the comments that we could remove that. Still think it's a bit weird though.

**Comment by alskipp** [comment]
I wouldn't consider it the perfect solution, but I think it's slightly different to option 2 above, as it requires the argument label when passing an argument, but allows `()` when no args are needed. Though, it would be preferable if the `()` wasn't required.

**Comment by tonyd256** [comment]
I think #220 helps with this issue since you don't need to call `.value` on a decoding if you don't want a `Decoded` type.

**Comment by sidraval** [comment]
@tonyd256 Can this be closed now that #238 is merged?

**Comment by tonyd256** [comment]
I think so
